spatialdata/_core/transform.py

```diff
@@ -1,86 +1,526 @@
+        m: ArrayLike = np.vstack((np.hstack((np.diag(self.scale), np.zeros((2, 1)))), np.array([0.0, 0.0, 1.0])))
+        return Affine(affine=m)
 
-        if scale_factors is None:
-            if translation is not None:
-                scale_factors = np.ones(len(translation), dtype=float)
-            elif ndim is not None:
-                scale_factors = np.ones(ndim, dtype=float)
+
+class Affine(BaseTransformation):
+    def __init__(
+        self,
+        affine: Optional[Union[ArrayLike, List[Any]]] = None,
+    ) -> None:
+        """
+        class for storing scale transformations.
+        """
+        if affine is None:
+            affine = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]], dtype=float)
+        else:
+            if isinstance(affine, list):
+                affine = np.array(affine)
+            if len(affine.shape) == 1:
+                assert len(affine) == 6
+                affine = np.vstack([affine.reshape((2, 3)), np.array([0.0, 0.0, 1.0])])
+            assert affine.shape == (3, 3)
+
+        self.affine = affine
+        self._ndim = 2
+
+    @property
+    def ndim(self) -> Optional[int]:
+        return self._ndim
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "type": "affine",
+            "affine": self.affine[:2, :].ravel().tolist(),
+        }
+
+    def transform_points(self, points: ArrayLike) -> ArrayLike:
+        p = np.vstack([points.T, np.ones(points.shape[0])])
+        q = self.affine @ p
+        return q[:2, :].T  # type: ignore[no-any-return]
+
+    def inverse(self) -> BaseTransformation:
+        # naive check for numerical stability
+        a = self.affine[0, 0]
+        b = self.affine[0, 1]
+        m = self.affine[0, 2]
+        c = self.affine[1, 0]
+        d = self.affine[1, 1]
+        n = self.affine[1, 2]
+        det = a * d - b * c
+        closed_form = np.array([[d, -c, 0], [-b, a, 0], [b * n - d * m, c * m - a * n, det]])
+        return Affine(affine=closed_form)
+
+    def to_affine(self) -> Affine:
+        return copy.deepcopy(self)
+
+
+class Rotation(BaseTransformation):
+    def __init__(
+        self,
+        rotation: Optional[Union[ArrayLike, List[Any]]] = None,
+    ) -> None:
+        """
+        class for storing scale transformations.
+        """
+        if rotation is None:
+            rotation = np.array([[1, 0], [0, 1]], dtype=float)
+        else:
+            if isinstance(rotation, list):
+                rotation = np.array(rotation)
+            if len(rotation.shape) == 1:
+                assert len(rotation) == 4
+                rotation = rotation.reshape((2, 2))
+            assert rotation.shape == (2, 2)
+
+        self.rotation = rotation
+        self._ndim = 2
+
+    @property
+    def ndim(self) -> Optional[int]:
+        return self._ndim
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "type": "rotation",
+            "rotation": self.rotation.ravel().tolist(),
```

Comment:
sometimes it is `.tolist()` some other it is `.ravel().tolist()` would converge in one.

Reply:
This is due to the fact that the specs sometimes store the matrix as a list and not a list of lists, and without the last row of the affine matrix, which is always [0, 0, 0, ..., 0, 1]. I will replace every class to accept as parameter to `__init__`, and store, a real affine matrix, and leave the job of removing the last row, and eventually flattening the list, to another separate part of the code.